src/pczt/tx_extractor.rs

```diff
@@ -0,0 +1,156 @@
+                    action
+                        .spend
+                        .spend_auth_sig
+                        .clone()
+                        .ok_or(TxExtractorError::MissingSpendAuthSig)?,
+                ))
+            },
+            |bundle| {
+                Ok(Unbound {
+                    proof: bundle
+                        .zkproof
+                        .clone()
+                        .ok_or(TxExtractorError::MissingProof)?,
+                    bsk: bundle
+                        .bsk
+                        .clone()
+                        .ok_or(TxExtractorError::MissingBindingSignatureSigningKey)?,
+                })
+            },
+        )
+    }
+
+    /// Converts this PCZT bundle into a regular bundle with the given authorizations.
+    fn to_tx_data<A, V, E, F, G>(
+        &self,
+        action_auth: F,
+        bundle_auth: G,
+    ) -> Result<Option<crate::Bundle<A, V>>, E>
+    where
+        A: Authorization,
+        E: From<TxExtractorError>,
+        F: Fn(&Action) -> Result<<A as Authorization>::SpendAuth, E>,
+        G: FnOnce(&Self) -> Result<A, E>,
+        V: TryFrom<i64>,
+    {
+        let actions = self
+            .actions
+            .iter()
+            .map(|action| {
+                let authorization = action_auth(action)?;
+
+                Ok(crate::Action::from_parts(
+                    action.spend.nullifier,
+                    action.spend.rk.clone(),
+                    action.output.cmx,
+                    action.output.encrypted_note.clone(),
+                    action.cv_net.clone(),
+                    authorization,
+                ))
+            })
+            .collect::<Result<_, E>>()?;
+
+        Ok(if let Some(actions) = NonEmpty::from_vec(actions) {
+            let value_balance = i64::try_from(self.value_sum)
+                .ok()
+                .and_then(|v| v.try_into().ok())
+                .ok_or(TxExtractorError::ValueSumOutOfRange)?;
+
+            let authorization = bundle_auth(&self)?;
+
+            Some(crate::Bundle::from_parts(
+                actions,
+                self.flags,
+                value_balance,
+                self.anchor,
+                authorization,
+            ))
+        } else {
+            None
+        })
+    }
+}
+
+/// Errors that can occur while extracting a regular Orchard bundle from a PCZT bundle.
+#[derive(Debug)]
+pub enum TxExtractorError {
+    /// The Transaction Extractor role requires `bsk` to be set.
+    MissingBindingSignatureSigningKey,
+    /// The Transaction Extractor role requires `zkproof` to be set.
+    MissingProof,
+    /// The Transaction Extractor role requires all `zkproof` fields to be set.
+    MissingSpendAuthSig,
+    /// The value sum does not fit into a `valueBalance`.
+    ValueSumOutOfRange,
+}
+
+/// Authorizing data for a bundle of actions that is just missing a binding signature.
+#[derive(Debug)]
+pub struct Unbound {
```

Comment:
nit: this name isn't particularly intuitive to me, but I don't have a better suggestion; maybe `BindingSigInputs` or something?

Reply:
The name appears primarily in type signatures, i.e. `Bundle<Unbound, ZatBalance>`; I chose it to make sense there.